### ISSUE TITLE
Ceph: config: revisit cluster CRD defaults (paths, namespaces, ...)

### ISSUE THREAD

**Issue by liewegas** [rook/rook#1347]
Several things stood out when looking at the cluster CRD:

- the default path, if not specified, makes rook use emptyDir.  defaulting to an emphemeral clsuter seems like a very dangerous/bad default!
- /var/lib/ceph/$namespace for the default instead of /var/lib/rook?  makes it very confusing (esp inside containers) when things are in default locations.
- drop cluster name inside ceph container?

**Comment by kokhang** [comment]
@liewegas one thing Rook is doing with the cluster CRD is to avoid exposing information about the underlying storage backend. Even though ceph is currently the only storage backend, I would like to make the Rook API (CRD) generic enough so that more storage backends can be added in the future.

You are right about the need to revisit the CRD. How about having a `ceph` element under `spec` for all the ceph specific params. We can even set the dataDir here to be `/var/lib/ceph`. For example:

```
apiVersion: rook.io/v1alpha1
kind: Cluster
metadata:
  name: rook
  namespace: rook
spec:
  dataDirHostPath: /var/lib/rook
  # toggle to use hostNetwork
  hostNetwork: false
  ceph:
    dataDir: /var/lib/ceph
    # set the amount of mons to be started
    monCount: 3
    storage: # cluster level storage configuration and selection
      useAllNodes: true
      useAllDevices: false
      deviceFilter:
      metadataDevice:
      location:
      storeConfig:
        storeType: bluestore
        databaseSizeMB: 1024 # this value can be removed for environments with normal sized disks (100 GB or larger)
        journalSizeMB: 1024  # this value can be removed for environments with normal sized disks (20 GB or larger)
```

**Comment by liewegas** [comment]
Yeah that makes sense.

Does hostNetwork affect rook, or just ceph?  Seems like it might just be ceph?

Ceph's dataDir should probably be dataDirHostPath?  There is no reason to make it anything else inside the container.  Is this the host path it binds to (e.g., with a namespace subdirectory inside it, then bind mounted to the container's /var/lib/ceph)?

**Comment by travisn** [comment]
@liewegas  here are my thoughts on these questions (somehow I missed this while on holiday):
- Some refactoring of the cluster CRD is being considered for #1375. This would make it clear which properties are specific to ceph and which would be needed by multiple backends. Rook itself doesn't need any settings in the cluster CRD since it's all about orchestrating the backend(s).
- hostNetwork does currently only affect the ceph daemons that are started by rook. In the future, if daemons are also started for another backend they would also need the hostNetwork setting. From that perspective, hostNetwork likely belongs outside the `ceph` element.
- Agreed that we should rework the paths inside the container to make it more natural to troubleshoot when connecting to it. The ceph pods could be bind mounted to /var/lib/ceph, while other pods in the future could be bind mounted to their default path. This also should mean we can simplify the ceph commands so they find the config in the default location instead of a custom path. The exception is that in the operator pod, it will use custom paths since it could be managing multiple storage clusters.

**Comment by jbw976** [comment]
@kfox1111 resources in Kubernetes are fully qualified with an API group and version also.  For example, Kind=Cluster is actually fully named something like `cluster.rook.io/v1alpha1`, which prevents any conflict with other types of cluster objects that may be defined as extensions to k8s.  Does that address your concern?

**Comment by kfox1111** [comment]
not sure. I'm use to seeing the names on the cli.
how does it work when you have cluster.rook.io/xxxxx kind=cluster, and watheverelse.io/xxx kind=cluster.

how does:
kubectl get cluster behave?

Thanks,
Kevin

**Comment by jbw976** [comment]
It looks like this comment addresses that behavior, but not sure if it's still the case in current day `kubectl`: https://github.com/kubernetes/kubernetes/issues/42521#issuecomment-284164886
> This issue is about how kubectl resolves an ambiguous resource name if no group is provided. The agreed on approach is to resolve to v1 resources first, then resolve in order of discovery information published by the server.

**Comment by kfox1111** [comment]
Yeah. Thats what I'm wondering about. Maybe k8s should probably decide the expected behaviour, and then we set the kind= based on that recommendation?

Just worried about writing a bunch of scripts that use kubectl get cluster and then when some user adds something else, those scripts break in really weird ways.

**Comment by jbw976** [comment]
Could your script just use the fully qualified name?  `kubectl -n rook get cluster.rook.io` should work reliably.

**Comment by travisn** [comment]
i believe this is addressed by #2230

**Comment by travisn** [comment]
This is already implemented for mons and mgr, and other daemons in progress by @BlaineEXE
